web/src/message_view.js

```diff
@@ -219,6 +219,46 @@ function handle_post_message_list_change(
     compose_recipient.handle_middle_pane_transition();
 }
 
+function try_rendering_locally_for_same_narrow(filter, opts) {
+    if (opts.then_select_id || opts.then_select_offset) {
+        // This function is designed to navigate user to a target message
+        // specified via `near` operator, which doesn't handle
+        // `then_select_id` or `then_select_offset` to avoid targeting
+        // the wrong message.
+        return false;
+    }
+
+    const current_filter = narrow_state.filter();
+    if (!current_filter) {
+        return false;
+    }
+
+    // If the difference between the current filter and the new filter
+    // is just a `near` operator, or just the value of a `near` operator,
+    // we can render the new filter without a rerender of the message list
+    // if the target message in the `near` operator is already rendered.
+    const excluded_operators = ["near"];
+    if (!filter.equals(current_filter, excluded_operators)) {
+        return false;
+    }
+
+    if (filter.has_operator("near")) {
+        const target_id = Number.parseInt(filter.operands("near")[0], 10);
+        const currently_selected_id = message_lists.current?.selected_id();
+        if (currently_selected_id !== target_id && message_lists.current?.get(target_id)) {
```

Comment:
Hmm, I see why the `message_lists.current?.get(target_id)` check is here to ensure we only do the scroll when the target is actually there ... but if the message is not present in the message feed, I'm not sure the fallthrough is correct; don't you want to `return false` otherwise?

Reply:
Nice catch! Fixed.